### DISCUSSION TITLE
Always On AP Mode/Manager Mode

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#19009] [Show and tell]
Separating out the discussion from #13362 so it's a little more visible for people searching for a hacky solution:

With an edited tasmota/tasmota_xdrv_driver/xdrv_01_9_webserver.ino (attached for 13.0.0.1)
[xdrv_01_9_webserver.ino.zip](https://github.com/arendst/Tasmota/files/11925494/xdrv_01_9_webserver.ino.zip), Tasmota can function in AP mode full-time, to allow management of a device with no access to wifi. This is useful for certain remote devices, like sensors, vehicle trackers, irrigation switches, AC IR controllers inside metal shipping containers or battery chargers in the middle of nowhere (my use-case is rural Africa where wifi is mostly non-existent except for people's phone hotspots).

There are certain considerations, of course. Time-keeping is usually important (but not always). $2 DS3231 modules work well for that, and are very accurate over long periods. Cheap $4 GPS modules also work well, but are kind of overkill.

Need to have 
```
#define USE_ALWAYS_AP
#define WIFI_CONFIG_TOOL       WIFI_MANAGER
#define WIFI_SCAN_AT_RESTART   false
#define WIFI_SCAN_REGULARLY    false
```
Should have
```
#define WIFI_AP_PASSPHRASE     ""
```
in user_config_override.h

Tested on both ESP8266 and ESP32. With `#define USE_ALWAYS_AP` commented out, the webserver driver functions normally.

A nice feature is that you can enter a wifi network in the wifi section, and the unit will search for it on startup for 20 seconds, then activate AP mode if not found. Thus, you can restart the unit from the AP mode, then activate your phone's hotspot, and the unit will join it, giving it network access to update the RTC, send MQTT messages, and any other connected features.  (I, for example, activate MQTT in AP mode, which triggers a restart, then the unit joins the phone hotspot, allowing remote control for troubleshooting). 

Simply turning off the phone hotspot will cause the unit to search for 20 more seconds then re-activate AP mode. Thus both modes can be entered without physical access to the unit.

Here's a generic build with a number of sensor drivers and rules. I usually build with scripting/SD card, since data logging without wifi requires that.
[tasmota.bin.gz](https://github.com/arendst/Tasmota/files/11925595/tasmota.bin.gz)

**Comment by kub3let** [comment]
I rebased against 13.4.0 and built images for more targets (ESP8266/ESP32/ESP32c3/ESP32s2)

- Flash correct image onto esp
  -  factory binary via usb/uart using https://tasmota.github.io/install
  - normal binary as ota upgrade via web gui on existing tasmota devices
- Connect to Tasmota AP
- Visit http://tasmota.local

**Firmware download:**
[tasmota-ap-13.4.0.zip](https://github.com/arendst/Tasmota/files/15199810/tasmota-ap-13.4.0.zip)

**Reply by martin072** [reply]
Nice! wouldn't it be an option for a feature request? or has that already been done?

**Reply by pkkrusty** [reply]
Has been discussed multiple times in the past. Main devs aren't interested in making it a feature. Introduces a fair amount of limitations that need to be worked around (need a RTC module to keep good time). But this hack serves my purposes quite well. I use it for remote watering stations run off solar.

**Reply by hawa-lc4** [reply]
Many Thanks to @pkkrusty for sharing this nice and helpful feature with the community.
It is exactly what I need for a light control out in the meadow securing switches with an RFID reader and time is held by RTC module.

@kub3let also Thanks to you for doing the rebase to 13.4.0 which I appreciate very much.
Would you be able the share the modified xdrv_01_9_webserver.ino file here as well?
I'll need the source file for I have to compile the complete bundle to fit my needs.
Would be kind, Thanks.
(I'm a little bit lazy and otherwise I would have to do this work again as you already did. ;)

**Reply by pkkrusty** [reply]
I think on the referenced thread, I linked the webserver ino file for a few different versions. I've used it unmodified up to 14.2, so probably will also work with the latest release. 

Edit: relevant comment: https://github.com/arendst/Tasmota/discussions/13362#discussioncomment-10367796

**Reply by hawa-lc4** [reply]
Thank you very much for your quick response.
On the first glance comparing the original file from 14.3.0.1 and the modified one it seems that the modifications can be taken to that version directly.
I'll do that and compile the bundle. Let's see what happens. I'll come back here after that.

**Reply by nikito7** [reply]
And permanent 60ma to 100ma power consumption.

With a switch it's a good option

**Reply by FulvioSpelta** [reply]
My test (already reported in issue #13362):

Downloaded 14.3.0 sources
Overwritten [xdrv_01_9_webserver.ino](https://github.com/user-attachments/files/16645718/xdrv_01_9_webserver.ino.zip)
Added the #define in "user_config_override.h" 
Compiled and flashed into a Sonoff Basic R2
It is running ok with AP always on and webadmin interface.
Thanks

**Reply by paggps** [reply]
> My test (already reported in issue #13362):
> 
> Downloaded 14.3.0 sources Overwritten [xdrv_01_9_webserver.ino](https://github.com/user-attachments/files/16645718/xdrv_01_9_webserver.ino.zip) Added the #define in "user_config_override.h" Compiled and flashed into a Sonoff Basic R2 It is running ok with AP always on and webadmin interface. Thanks

Hello @FulvioSpelta. Can you please post the binary of Tasmota-ap compiled from 14.3.0 sources (or newer)?
Thank you very much.
JJ

**Reply by FulvioSpelta** [reply]
@paggps attached.

```
[firmware.zip](https://github.com/user-attachments/files/18048161/firmware.zip)
```

Note that the AP will be open. To set the passphrase you have to compile by your own.

`#define USE_ALWAYS_AP
#define WIFI_CONFIG_TOOL       WIFI_MANAGER        // [WifiConfig] Default tool if Wi-Fi fails to connect (default option: 4 - 
#define WIFI_SCAN_AT_RESTART   false             // [SetOption56] Scan Wi-Fi network at restart for configured AP's
#define WIFI_SCAN_REGULARLY    false              // [SetOption57] Scan Wi-Fi network every 44 minutes for configured AP's
#define WIFI_AP_PASSPHRASE     ""                // AccessPoint passphrase. For WPA2 min 8 char, for open use "" (max 63 char).`

`Processing tasmota (platform: https://github.com/tasmota/platform-espressif8266/releases/download/2024.09.00/platform-espressif8266.zip; framework: arduino; board: esp8266_1M)

CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/esp8266_1M.html   
PLATFORM: Espressif 8266 (2024.9.0) > Espressif Generic ESP8266 Tasmota 1M sketch NO FS
HARDWARE: ESP8266 80MHz, 80KB RAM, 972.00KB Flash`

**Reply by paggps** [reply]
Thank you very much @FulvioSpelta  ! Just flashed and worked ok !

**Reply by GEAG** [reply]
> And permanent 60ma to 100ma power consumption.

Does anybody have an idea, if it would be possible to reduce the power consumption of AP-mode (modemsleep?) by at least -25%, better -40% ?

**Reply by sfromis** [reply]
I'd say that modem sleep and AP mode does not seem like a good match, but Tasmota has no support for modem sleep, only deep sleep.

**Comment by hawa-lc4** [comment]
@pkkrusty the xdrv_01_9_webserver.ino files of 14.2 and 14.3 seem to be absolutely the same. I've found no differences except the modifications done regarding this feature. So I've placed your file directly to the folder.
Compilation without errors, resulting .bin files works like a charm.
I've tested only a few things like pure AP-mode, adding optional WiFi credentials to access a temporary AP on smartphone: all work fine.

@nikito7 power consumption is always a special topic for those devices.
However as long as you are not on battery it's only of lower interest.
The AC/DC power supplies of that kind have mostly a really bad efficiency.
The mains relay controlled by this circuit here switches lights of nearly 1000 Watt, so no relationship to 1 Watt consumption of the MCU.

**Reply by axel-GB** [reply]
What about config WIFI_AP_PASSPHRASE to set or change from tasmota GUI  ??

**Comment by axel-GB** [comment]
It's not the same as Gui. I'm talking about changing (saving) from Gui
without flashing with new password

pon., 27 sty 2025, 15:43 użytkownik Fulvio Spelta ***@***.***>
napisał:

**Reply by FulvioSpelta** [reply]
I think that this capability has to be added.
Maybe that the current support for STA mode wifi key can be "duplicated" to support the AP mode key.
I can't estimate the effort needed.

**Reply by nikito7** [reply]
Anyone tested:

BackLog Wifi 0; Wifi 1

If it work ok?

**Reply by nikito7** [reply]
Do something here:

https://github.com/arendst/Tasmota/blob/9613004bb42a1b307b8c0c6ccacf0673630dd120/tasmota/tasmota_xdrv_driver/xdrv_01_9_webserver.ino#L718

```c
  // bool softAP(const char* ssid, const char* passphrase = NULL, int channel = 1, int ssid_hidden = 0, int max_connection = 4);
  WiFi.softAP(TasmotaGlobal.hostname, WIFI_AP_PASSPHRASE, channel, 0, 1);
```

**Comment by nikito7** [comment]
Just a diff...

```diff
diff --git a/tasmota/ap-14.3.0.ino b/tasmota/ap-14.3.0.ino
index b9837b78..471817e1 100644
--- a/tasmota/ap-14.3.0.ino
+++ b/tasmota/ap-14.3.0.ino
@@ -523,7 +523,11 @@ String AddWebCommand(const char* command, const char* arg, const char* dflt) {
 }
 
 static bool WifiIsInManagerMode(){
-  return (HTTP_MANAGER == Web.state || HTTP_MANAGER_RESET_ONLY == Web.state);
+  #ifndef USE_ALWAYS_AP
+  return (HTTP_MANAGER == Web.state || HTTP_MANAGER_RESET_ONLY == Web.state); //ADJUSTMENT DEFAULT
+  #else
+  return false; //ADJUSTMENT
+  #endif
 }
 
 void ShowWebSource(uint32_t source)
@@ -621,10 +625,14 @@ void StartWebserver(int type)
       XdrvXsnsCall(FUNC_WEB_ADD_HANDLER);
 #endif  // Not FIRMWARE_MINIMAL
 
-      if (!Web.initial_config) {
-        Web.initial_config = (!strlen(SettingsText(SET_STASSID1)) && !strlen(SettingsText(SET_STASSID2)));
-        if (Web.initial_config) { AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_HTTP "Blank Device - Initial Configuration")); }
+#ifndef USE_ALWAYS_AP
+      if (!Web.initial_config) {  //ADJUSTMENT DEFAULT
+        Web.initial_config = (!strlen(SettingsText(SET_STASSID1)) && !strlen(SettingsText(SET_STASSID2)));    //ADJUSTMENT DEFAULT
+        if (Web.initial_config) { AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_HTTP "Blank Device - Initial Configuration")); }    //ADJUSTMENT DEFAULT
       }
+#else
+      Web.initial_config = true; //ADJUSTMENT to keep the unit from restarting every 3 minutes if wifi credentials are entered
+#endif
     }
     Web.reset_web_log_flag = false;
 
@@ -673,7 +681,12 @@ void WifiManagerBegin(bool reset_only)
   DnsServer->setErrorReplyCode(DNSReplyCode::NoError);
   DnsServer->start(DNS_PORT, "*", WiFi.softAPIP());
 
-  StartWebserver((reset_only ? HTTP_MANAGER_RESET_ONLY : HTTP_MANAGER));
+#ifndef USE_ALWAYS_AP
+  StartWebserver((reset_only ? HTTP_MANAGER_RESET_ONLY : HTTP_MANAGER)); //ADJUSTMENT DEFAULT
+#else
+  Web.state = HTTP_OFF; //ADJUSTMENT
+  StartWebserver((reset_only ? HTTP_MANAGER_RESET_ONLY : HTTP_ADMIN)); //ADJUSTMENT
+#endif
 }
 
 void PollDnsWebserver(void)
@@ -1108,6 +1121,10 @@ void WebRestart(uint32_t type) {
 #endif  // ESP32
     TasmotaGlobal.restart_flag = 2;
   }
+  #ifdef USE_ALWAYS_AP
+  Web.state = HTTP_ADMIN; //ADJUSTMENT
+  WSContentSpaceButton(BUTTON_MAIN); //ADJUSTMENT
+  #endif
   WSContentStop();
 }
 
```

**Reply by nikito7** [reply]
This could be interesting with LoRa

**Reply by FulvioSpelta** [reply]
Sorry I don't understand clearly: the file that has been changed to have always on AP is [xdrv_01_9_webserver.ino](https://github.com/user-attachments/files/16645718/xdrv_01_9_webserver.ino.zip)
Where the diff posted came from ? Thanks

**Reply by nikito7** [reply]
It's the same. I just changed name

**Reply by axel-GB** [reply]
And what does the diff do ? AP Password in Gui ?

**Reply by nikito7** [reply]
> And what does the diff do ? AP Password in Gui ?

It's just the old code. Just to see. Nothing new...

**Comment by axel-GB** [comment]
How can i use AP2 password like this
// bool softAP(const char* ssid, const char* passphrase = NULL, int channel = 1, int ssid_hidden = 0, int max_connection = 4);
WiFi.softAP(TasmotaGlobal.hostname, **HERE AP2 PASSWORD**, channel, 0, 1);

**Comment by ramponis** [comment]
Do you think it is possibile to use this ack on this device?

https://templates.blakadder.com/lilygo_T-Internet-POE.html

My purpose is to connect the device using ethernet and use wifi in ap mode.
So, it will become a really access point.
Do you think it will be possible?

**Reply by pkkrusty** [reply]
Seems like more trouble than it's worth. Throughput will be quite limited. Makes more sense to me to buy an old router for $5 and everything works with no hacks.

**Reply by ramponis** [reply]
Yes, you are right but i have a complicated situation...
I cannot give power to the device, so I will use the poe.
This device will be used to interface some sensors.
Also I want to use it in AP mode to permit connection to another tasmota device not covered by wifi.
So the throughput will not be very important

**Reply by sfromis** [reply]
No, Tasmota does not allow you to create a real AP with Ethernet backhaul. Closest is https://tasmota.github.io/docs/Range-Extender/ where you can have an AP, and Wifi backhaul.

**Reply by ramponis** [reply]
Thank you, this can be enought for my pourpose

**Comment by Jayeshkarthick** [comment]
Hello, I'm new to Tasmota, I just want sonoff basic r4 to work in AP mode, I'll connect my android phone to its wifi and send http based commands. No other requirement

While i try to compile with VSCODE and Platform IO according to above changes you guys mentioned i can create the firmware but while implementing in Sonoff it just doesn't do anything - no hotspot. 

Attaching megalink for my files: https://mega.nz/file/AiQFnJjZ#XV1766dE5oEVbzjb37DbHT4qEJLW0pLCXFN1kA4atHk

Any insights would help a lot

**Reply by pkkrusty** [reply]
Can you hook up your esp device to usb and watch on minicom on what's going on with the serial log?
